usr/sbin/p11sak/p11sak.c

```diff
@@ -1142,6 +1192,152 @@ static CK_ATTRIBUTE_TYPE col2type(int col)
+                type = confignode_find(structnode->value, "type");
+
+                // checking syntax of attribute...
+                if (name == NULL) {
+                    fprintf(stderr, "Missing name in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                } else if (!confignode_hastype(name, CT_BAREVAL)) {
+                    fprintf(stderr, "Invalid name in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                } else if (hex_string == NULL) {
+                    fprintf(stderr, "Missing value in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                } else if (!confignode_hastype(hex_string, CT_INTVAL)) {
+                    fprintf(stderr, "Invalid name in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                } else if (type == NULL) {
+                    fprintf(stderr, "Missing type in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                } else if (!confignode_hastype(type, CT_BAREVAL)) {
+                    fprintf(stderr, "Invalid name in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                }
+                
+                // print attribute by type
+                unsigned int hex = confignode_to_intval(hex_string)->value;
+                if (strcmp((confignode_to_bareval(type)->value), "CK_BBOOL") == 0) {
+                    // build template
+                    CK_BBOOL a_bool;
+                    CK_ULONG bs = sizeof(CK_BBOOL);
+                    CK_ATTRIBUTE temp = {hex, &a_bool, bs};
+
+                    // get attribute value
+                    rc = funcs->C_GetAttributeValue(session, hkey, &temp, 1);
+                    if (rc == CKR_OK) {
+                        if (temp.ulValueLen != sizeof(CK_BBOOL)) {
+                            printf(" Error in retrieving Attribute %s: %lu\n",
+                                    confignode_to_bareval(name)->value, temp.ulValueLen);
+                        } else {
+                            if (long_print) {
+                                printf("          %s: %s\n", confignode_to_bareval(name)->value,
+                                        CK_BBOOL2a(*(CK_BBOOL*) temp.pValue));
+                            } else if (strcmp(CK_BBOOL2a(*(CK_BBOOL*) temp.pValue), "CK_TRUE") == 0) {
+                                printf(" 1 ");
+                                return rc;
+                            }
+                        }    
+                    } else if (rc == CKR_ATTRIBUTE_SENSITIVE) {
+                        printf("          %s: SENSITIVE\n", confignode_to_bareval(name)->value);
+                    } else if (rc != CKR_ATTRIBUTE_TYPE_INVALID) {
+                        return rc;
+                    } 
```

Comment:
You should also have a final else case, if rc is not CKR_OK, CKR_ATTRIBUTE_SENSITIVE, nor CKR_ATTRIBUTE_TYPE_INVALID, and thread that (unexpected) return code as an error.

Reply:
Its OK as is, check is `rc != CKR_ATTRIBUTE_TYPE_INVALID`

Reply:
The "final" else case is here the else if (rc != CKR_ATTRIBUTE_TYPE_INVALID), the Error is returned and in Case rc == CKR_ATTRIBUTE_TYPE_INVALID then the attribute is skipped.